src/order.c

```diff
@@ -0,0 +1,4148 @@
+                          struct lazy_vec* p_lazy_counts,
+                          struct group_infos* p_group_infos,
+                          bool decreasing,
+                          bool na_last,
+                          R_xlen_t size) {
+  if (size <= INSERTION_ORDER_BOUNDARY) {
+    int_adjust(p_x, decreasing, na_last, size);
+    int_insertion_order(p_x, p_o, p_group_infos, size);
+    return;
+  }
+
+  lazy_vec_initialize(p_lazy_o_aux);
+  int* p_o_aux = (int*) p_lazy_o_aux->p_data;
+
+  uint32_t range;
+  int x_min;
+
+  int_compute_range(p_x, size, &x_min, &range);
+
+  /*
+   * If in counting order range and on the second or higher column, we will
+   * need `p_o_aux` as working memory. At this point, `p_o` will have been
+   * initialized from ordering the first column.
+   */
+  if (range < INT_COUNTING_ORDER_RANGE_BOUNDARY) {
+    const bool initialized = true;
+
+    int_counting_order(
+      p_x,
+      p_o,
+      p_o_aux,
+      p_group_infos,
+      size,
+      x_min,
+      range,
+      initialized,
+      decreasing,
+      na_last
+    );
+
+    return;
+  }
+
+  lazy_vec_initialize(p_lazy_x_aux);
+  uint32_t* p_x_aux = (uint32_t*) p_lazy_x_aux->p_data;
+
+  lazy_vec_initialize(p_lazy_bytes);
+  uint8_t* p_bytes = (uint8_t*) p_lazy_bytes->p_data;
+
+  lazy_vec_initialize(p_lazy_counts);
+  R_xlen_t* p_counts = (R_xlen_t*) p_lazy_counts->p_data;
+  memset(p_counts, 0, p_lazy_counts->size);
+
+  int_adjust(p_x, decreasing, na_last, size);
+
+  int_radix_order(
+    p_x,
+    p_o,
+    p_x_aux,
+    p_o_aux,
+    p_bytes,
+    p_counts,
+    p_group_infos,
+    size
+  );
+}
+
+static inline void* int_maybe_copy(const int* p_x,
+                                   struct lazy_vec* p_lazy_x_chunk,
+                                   R_xlen_t size,
+                                   bool copy);
+
+/*
+ * `int_order_impl()` is used by both `int_order()` and by
+ * `chr_order()`.
+ *
+ * For `chr_order()`, the TRUELENGTH values will already be extracted into
+ * `p_lazy_x_chunk`. In this case, we set `copy = false` to tell
+ * `int_order_impl()` to use `p_lazy_x_chunk` directly rather than copying `p_x`
+ * over to `p_x_chunk`.
+ */
+static
+void int_order_impl(const int* p_x,
+                    struct lazy_order* p_lazy_o,
+                    struct lazy_vec* p_lazy_x_chunk,
+                    struct lazy_vec* p_lazy_x_aux,
+                    struct lazy_vec* p_lazy_o_aux,
+                    struct lazy_vec* p_lazy_bytes,
+                    struct lazy_vec* p_lazy_counts,
+                    struct group_infos* p_group_infos,
+                    bool decreasing,
+                    bool na_last,
+                    R_xlen_t size,
+                    bool copy) {
+  if (size <= INSERTION_ORDER_BOUNDARY) {
+    lazy_order_initialize(p_lazy_o);
+    int* p_o = p_lazy_o->p_o;
+
+    void* p_x_chunk = int_maybe_copy(p_x, p_lazy_x_chunk, size, copy);
+    int_adjust(p_x_chunk, decreasing, na_last, size);
```

Comment:
Two reasons:

1) At this specific location in the code, it is more about handling `na_last` and `decreasing` ahead of time before handing off to the insertion sort. Theoretically an insertion sort could handles ints fine. However...

2) This specific insertion sort code is also used in the radix sort to "finish off" small groups. The data there has already been adjusted to be unsigned so the insertion sort has to deal with unsigned input.

Rather than write another version of `int_adjust()` that just handles `na_last` and `decreasing` (and doesn't map to unsigned) and another version of `int_insertion_sort()` that works for signed ints, I figured it was easiest to just reuse the existing unsigned version